cs/cs/Serializer/VowpalWabbitJsonSerializer.cs

```diff
@@ -0,0 +1,298 @@
+
+        static VowpalWabbitJsonSerializer()
+        {
+            // find mapping from property names to types
+            var q = from t in new[] { typeof(SimpleLabel), typeof(ContextualBanditLabel) }
+                    from p in t.GetProperties()
+                    let jsonProperty = (JsonPropertyAttribute)p.GetCustomAttributes(typeof(JsonPropertyAttribute), true).FirstOrDefault()
+                    where jsonProperty != null
+                    select new
+                    {
+                        Type = t,
+                        JsonProperty = jsonProperty,
+                        Property = p
+                    };
+
+            labelPropertyMapping = q.ToDictionary(e => e.JsonProperty.PropertyName ?? e.Property.Name, e => e.Type);
+        }
+        private readonly VowpalWabbit vw;
+        private readonly VowpalWabbitDefaultMarshaller defaultMarshaller;
+        private readonly JsonSerializer jsonSerializer;
+        private JsonReader reader;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VowpalWabbitJson"/> class.
+        /// </summary>
+        /// <param name="vw">The VW native instance.</param>
+        public VowpalWabbitJsonSerializer(VowpalWabbit vw)
+        {
+            Contract.Requires(vw != null);
+
+            this.vw = vw;
+            this.defaultMarshaller = new VowpalWabbitDefaultMarshaller();
+            this.jsonSerializer = new JsonSerializer();
+        }
+
+        /// <summary>
+        /// Parses the example.
+        /// </summary>
+        /// <param name="json">The example to parse.</param>
+        /// <returns>The VowpalWabbit native example.</returns>
+        public VowpalWabbitExample Parse(string json)
+        {
+            using (var textReader = new JsonTextReader(new StringReader(json)))
+            {
+                return this.Parse(textReader);
+            }
+        }
+
+        /// <summary>
+        /// Parses the example.
+        /// </summary>
+        /// <param name="reader">The example to parse.</param>
+        /// <returns>The VowpalWabbit native example.</returns>
+        public VowpalWabbitExample Parse(JsonReader reader)
+        {
+            // avoid parameter passing for the sake of non-reentrantness
+            this.reader = reader;
+            using (VowpalWabbitMarshalContext context = new VowpalWabbitMarshalContext(this.vw))
+            using (VowpalWabbitMarshalContext defaultNamespaceContext = new VowpalWabbitMarshalContext(this.vw, context.ExampleBuilder))
+            {
+                if (!reader.Read())
+                    throw new ArgumentException("json");
+
+                if (reader.TokenType != JsonToken.StartObject)
+                    throw new ArgumentException("json");
+
+                Namespace defaultNamespace = new Namespace(this.vw);
+                using (defaultNamespaceContext.NamespaceBuilder = defaultNamespaceContext.ExampleBuilder.AddNamespace(' '))
```

Comment:
maybe this '  ' space and the "_" characters below can be refactored out to const variables together somewhere for easy reference & documentation